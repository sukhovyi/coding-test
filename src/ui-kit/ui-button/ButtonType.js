import { PrimaryButton, SecondaryButton, DangerButton, SuccessButton, TertiaryButton } from './StyledButtons';

const ButtonType = Object.freeze({
  PRIMARY: 'primary',
  DANGER: 'danger',
  SUCCESS: 'success',
  SECONDARY: 'secondary',
  TERTIARY: 'tertiary',

  getButtons() {
    return {
      PrimaryButton,
      SecondaryButton,
      DangerButton,
      SuccessButton,
      TertiaryButton,
    };
  },

  getButtonByType(type) {
    const button = {
      [ButtonType.PRIMARY]: PrimaryButton,
      [ButtonType.DANGER]: DangerButton,
      [ButtonType.SUCCESS]: SuccessButton,
      [ButtonType.SECONDARY]: SecondaryButton,
      [ButtonType.TERTIARY]: TertiaryButton,
    }[type];

    if (!button) {
      throw new Error(`Button with type ${type} does not exist!`);
    }

    return button;
  },
});

export default ButtonType;
