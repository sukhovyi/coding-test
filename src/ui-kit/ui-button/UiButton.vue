<template>
  <component :is="buttonComponent" @click="onClick">
    <slot/>
  </component>
</template>

<script>
import ButtonType from './ButtonType';

export default {
  name: 'UiButton',
  components: {
    ...ButtonType.getButtons(),
  },
  props: {
    type: {
      type: String,
      default: ButtonType.PRIMARY,
    },
  },
  computed: {
    buttonComponent() {
      return ButtonType.getButtonByType(this.type);
    },
  },
  methods: {
    onClick() {
      this.$emit('click');
    },
  },
};
</script>
