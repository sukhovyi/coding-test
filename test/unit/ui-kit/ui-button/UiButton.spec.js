import { shallowMount } from '@vue/test-utils';
import UiButton from '../../../../src/ui-kit/ui-button/UiButton';
import ButtonType from '../../../../src/ui-kit/ui-button/ButtonType';

describe('ui-kit/ui-button/UiButton', () => {
  const wrapper = shallowMount(UiButton, {
    propsData: {
      type: ButtonType.SUCCESS,
    },
  });

  it('should be rendered correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('triggers a click', async () => {
    await wrapper.trigger('click');
    expect(wrapper.emitted().click).toBeTruthy();
  });
});
