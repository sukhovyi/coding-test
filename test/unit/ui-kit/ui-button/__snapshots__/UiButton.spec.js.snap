// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui-kit/ui-button/UiButton should be rendered correctly 1`] = `<styledbutton-stub />`;
